### PR TITLE
feat(wallet): add grpc method for setting base node

### LINKED ISSUE #3821: Create a gRPC interface in console wallet to set the connected base node
Address the following TODO in the code:

// TODO gRPC interfaces for setting base node 

clear TODO comment when completed

### PR DESCRIPTION
**Pull request by zhangcheng**
Description
---
Aim to resolve #3821

Motivation and Context
---
Learning via "good first issue"

How Has This Been Tested?
---
Code compiles, and manual test with `grpcurl`.

### CODE REVIEW

**Review by delta1 — COMMENTED** [review]
Thanks @zhangcheng looks good! You will just have to run `cargo fmt --all` to pass clippy CI. 

Regarding testing: 
- manually: use a gRPC client to to call against your method with the wallet running 
- automated: take a look at `integration_tests` folder (but you can leave this out for now)

**Comment by zhangcheng** [comment]
Got it, thx @delta1 , will clippy it.

**Review by philipr-za — COMMENTED** [review]
Looks good, will approve once Clippy passes.

**Comment by zhangcheng** [comment]
I tried to use grpcurl to test against it, but my limited knowledge couldn't help me to construct valid inputs.
It'd be very helpful if there is a brief guide or alike to educate newbies like me on how to properly construct needed inputs with the correct format.

```
grpcurl -plaintext \
    -import-path ./applications/tari_app_grpc/proto \
    -proto wallet.proto \
    -d '{"public_key": "36a9df45e1423b5315ffa7a91521924210c8e1d1537ad0968450f20f21e5200d", "net_address": "localhost"}' \
    localhost:18143 \
    tari.rpc.Wallet.SetBaseNode
```

**Comment by philipr-za** [comment]
> I tried to use grpcurl to test against it, but my limited knowledge couldn't help me to construct valid inputs. It'd be very helpful if there is a brief guide or alike to educate newbies like me on how to properly construct needed inputs with the correct format.
> 
> ```
> grpcurl -plaintext \
>     -import-path ./applications/tari_app_grpc/proto \
>     -proto wallet.proto \
>     -d '{"public_key": "36a9df45e1423b5315ffa7a91521924210c8e1d1537ad0968450f20f21e5200d", "net_address": "localhost"}' \
>     localhost:18143 \
>     tari.rpc.Wallet.SetBaseNode
> ```

I haven't personally tested a CLI gRPC client at all, I prefer a GUI client like BloomRPC which can just ingest the proto file. from a glance that request looks ok except for the format of the `net_address` field. We are using the self describing MultiAddr which would look like `/ip4/127.0.0.1`.

**Comment by delta1** [comment]
I agree with @philipr-za it's much easier to test with a GUI. But not a showstopper here, if the CI passes then LGTM

**Review by stringhandler — DISMISSED** [review]

**Comment by stringhandler** [comment]
The address can be `/ip4/127.0.0.1/tcp/18142`

**Comment by stringhandler** [comment]
Just need to sign your commits

**Comment by zhangcheng** [comment]
> ERROR:
>   Code: InvalidArgument
>   Message: Base node public key was not a valid pub key: The input data was the incorrect length to perform the desired conversion
> 

Since I don't know how to construct a valid public key string for this case yet, the above string was pulled out of the log files, such as:

> 2022-02-14 12:00:06.557504000 [comms::node] [Thread:4469540224] INFO  Your node's public key is '1ed8d6e659bdabc5bd1f7cb6f8cfbbdbd97f5413fb8c40fc35fcc4cccb54df69'

**Comment by stringhandler** [comment]
The grpc client often wants bytes to be base64, so you can try `HtjW5lm9q8W9H3y2+M+729l/VBP7jED8NfzEzMtU32k=`

**Comment by zhangcheng** [comment]
> Just need to sign your commits

Something new for me to learn then. 😄

**Comment by zhangcheng** [comment]
Cool, base64 string works. Thanks @mikethetike 🤝 
```
grpcurl -plaintext \
    -import-path ./applications/tari_app_grpc/proto \
    -proto wallet.proto \
    -d '{"public_key": "HtjW5lm9q8W9H3y2+M+729l/VBP7jED8NfzEzMtU32k=", "net_address": "/ip4/127.0.0.1/tcp/18142"}' \
    localhost:18143 \
    tari.rpc.Wallet.SetBaseNode
```

Console output:
```
Tari Console Wallet running... (gRPC mode started)
Setting base node peer...
1ed8d6e659bdabc5bd1f7cb6f8cfbbdbd97f5413fb8c40fc35fcc4cccb54df69::/ip4/127.0.0.1/tcp/18142
```

**Comment by delta1** [comment]
Well done 👏

**Review by delta1 — DISMISSED** [review]

**Comment by delta1** [comment]
@zhangcheng https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits

File: applications/tari_console_wallet/src/grpc/wallet_grpc_server.rs [hunk 1]
```diff
@@ -150,6 +152,29 @@ impl wallet_server::Wallet for WalletGrpcServer {
         }))
     }
 
+    async fn set_base_node(
+        &self,
+        request: Request<SetBaseNodeRequest>,
+    ) -> Result<Response<SetBaseNodeResponse>, Status> {
+        let message = request.into_inner();
+        let public_key = PublicKey::from_bytes(message.public_key.as_slice())
```
Comment (philipr-za) [inline applications/tari_console_wallet/src/grpc/wallet_grpc_server.rs]: Based on @mikethetike's comment I realized this is the problem. This should actually be using from_hex(...).

Comment (philipr-za) [reply applications/tari_console_wallet/src/grpc/wallet_grpc_server.rs]: but base64 bytes also does the job

Comment (stringhandler) [reply applications/tari_console_wallet/src/grpc/wallet_grpc_server.rs]: it would be if it were `string` and not `bytes` in the proto file

Comment (zhangcheng) [reply applications/tari_console_wallet/src/grpc/wallet_grpc_server.rs]: I see, will do it along with the commit signing. 😄

Comment (zhangcheng) [reply applications/tari_console_wallet/src/grpc/wallet_grpc_server.rs]: Changed, so test as:
```
grpcurl -plaintext \
    -import-path ./applications/tari_app_grpc/proto \
    -proto wallet.proto \
    -d '{"public_key_hex": "1ed8d6e659bdabc5bd1f7cb6f8cfbbdbd97f5413fb8c40fc35fcc4cccb54df69", "net_address": "/ip4/127.0.0.1/tcp/18142"}' \
    localhost:18143 \
    tari.rpc.Wallet.SetBaseNode
```

Also learned to sign the commit.

**Comment by zhangcheng** [comment]
> I haven't personally tested a CLI gRPC client at all, I prefer a GUI client like BloomRPC which can just ingest the proto file.

Just want to say, [BloomRPC](https://github.com/bloomrpc/bloomrpc) is much nicer and easier to use than [grpcurl](https://github.com/fullstorydev/grpcurl), esp. to a newbie. @philipr-za

**Review by stringhandler — APPROVED** [review]
I preferred the binary version, but this is fine too.

**Comment by stringhandler** [comment]
I've queued it. If the CI passes, it will get merged. Thanks so much!

**Pull request merged**